Add documentation about static pods.

jsafrane opened the pull request:
There is no description of static pods. Is it supported? If so, it should be described how it's supposed to work.

See it as html at https://github.com/jsafrane/kubernetes/blob/devel/static-pods/docs/static-pods.md

satnam6502 commented:
@dchen1107 : do we want to document static pods? Or perhaps not if we intend to get rid of them?

eparis commented:
@googlebot he signed the cla

childsb commented:
This feature is important for volume storage. We need a "utility" pod of some sort on each node  so that we can containerize the filesystem client connection.

I'd like to see it remain in kubernetes, or be replaced by similar. There's other use cases for a 'static' pod on every node in the cluster such as logging, monitoring/metering, etc..

rootfs commented on docs/static-pods.md:
```diff
@@ -0,0 +1,100 @@
+        $ ssh my-minion1
+
+2. Choose a directory, say `/etc/kubelet.d` and place a web server pod definition there, e.g. `/etc/kubernetes.d/static-web.yaml`:
+
```
would look better if you can add some syntax hightlight

satnam6502 commented:
@childsb I am quite familiar with the need to have a pod on each node. However, I wish I would do this with a per-node controller.

eparis commented:
or as @timstclair likes to think  SPREADBY(kubernetes.io/hostname)

bgrant0607 commented on docs/static-pods.md:
```diff
@@ -0,0 +1,100 @@
+Static pod can be created in two ways: either by using configuration file(s) or by HTTP.
+
+### Configuration files
+The configuration files are just standard pod definition in json or yaml format in specific directory. Use `kublet --config=<the directory>` to start kubelet daemon, which periodically scans the directory and creates/deletes static pods as yaml/json files appear/disappear there.
```
kubelet

jsafrane replied:
Fixed

bgrant0607 commented on docs/static-pods.md:
```diff
@@ -0,0 +1,100 @@
+
+        my-minion1$ mkdir /etc/kubernetes.d/
+        my-minion1$ cat <<EOF >/etc/kubernetes.d/static-web.yaml
+        apiVersion: v1beta3
```
Please convert to the v1 API.

jsafrane replied:
Fixed

bgrant0607 commented:
Thanks, but we plan to deprecate this mechanism in the not-too-distant future and replace it with a Daemon controller #1518. If we were going to document this feature, that should be mentioned prominently at the top, and we should move this doc out of the top-level docs directory, which should contain user-oriented docs. This is more a cluster admin feature, which we admittedly don't have a home for yet, other than getting-started-guides. Perhaps it could go in that subdir.

dchen1107 commented:
Yes, we are going to replace static pods with Daemon controller. I would rather not introduce such doc and make it as a hidden feature so that we can easily deprecated this later.

jsafrane commented:
This daemon controller feature is open for half a year. Is there any ETA? Until it's available static pods are probably the easiest way how to run a pod on all nodes (with proper scripting, e.g. using puppet or ansible).

bgrant0607 commented:
I hope we can merge the initial version of the Daemon controller shortly after our 1.0 freeze is over.

@AnanyaKumar @rjnagal @vmarmol

bgrant0607 commented:
Please comment on #10210. Perhaps we can even incorporate some of this into the alternative section of that doc.
cc @AnanyaKumar

erictune commented:
Don't we still need static pods for bootstrapping?

jsafrane commented:
My use case for static pods was to mount e.g. Gluster volume on nodes where mount.glusterfs is in a container - this pod/container needs to run on all hosts and could be static. This can be fully replaced by daemon controller. I don't know if anyone has any other use cases.

Anyway, as long as there is code for static pods, I think it would be nice to document it. Feel free to mark it as deprecated. Or remove the code.

bgrant0607 commented:
@erictune Bootstrapping: Today, yes. Eventually, no. I want Kubelet to expose an API endpoint on localhost to launch pods.

bgrant0607 commented:
It's @dchen1107's call about whether to document this feature. 

If we do, my requirements are:
1. Mark that it's deprecated in bold at the top.
2. Put in the getting-started directory.
3. Update to the v1 API

dchen1107 commented:
@jsafrane are you going to finish this pr, or close it? 

The initial pr for supporting daemon controller is merged lately too.

bgrant0607 commented on docs/static-pods.md:
```diff
@@ -0,0 +1,100 @@
+
+Labels from the static pod are propagated into the mirror-pod and can be used as usual for filtering.
+
+Notice we cannot delete the pod with the API server (e.g. via `kubeclt` command), kubelet simply won't remove it.
```
typo: kubectl

Should also link to the doc. Use relative link (from getting-started): ../user-guide/kubectl

bgrant0607 commented:
Also need to run hack/update-generated-docs.sh

jsafrane commented:
Sorry for late reply...

> If we do, my requirements are:
> 1. Mark that it's deprecated in bold at the top.
> 2. Put in the getting-started directory.
> 3. Update to the v1 API
> Also need to run hack/update-generated-docs.sh

All done, I am not sure about getting-started directory though. Wouldn't be `docs/admin` a better choice?

bgrant0607 commented:
I'm fine with docs/admin.

bgrant0607 commented on docs/getting-started-guides/static-pods.md:
```diff
@@ -0,0 +1,143 @@
+
+Running kubelet periodically scans the configured directory (`/etc/kubelet.d` in our example) for changes and adds/removes pods as files appear/disappear in this directory.
+
+    my-minion1$ mv /etc/kubernetes.d/static-web.yaml /tmp
```
Please use ```console

jsafrane committed:
Add documentation about static pods.

Described behavior was extracted from source code and via trial-and-error
attempts.

jsafrane commented:
* Moved to docs/admin
* Added ```console where appropriate (+ some editorial changes)

jsafrane commented:
> Builds failed on Shippable
>> Cloning into '/root/workspace/src/github.com/mattn/goveralls'...
fatal: unable to access 'https://github.com/mattn/goveralls/': Failed connect to github.com:443; Connection timed out


github seems to be in bad shape today...

dchen1107 commented on docs/admin/static-pods.md:
```diff
@@ -0,0 +1,163 @@
+
+# Static pods (deprecated)
+
+**Static pods are deprecated and can be removed in any future Kubernetes release!**
```
s/deprecated/to be deprecated?

dchen1107 replied:
Mention Daemon controller #1518 would be the one replacing static pod?

jsafrane replied:
> s/deprecated/to be deprecated?
done

> Mention Daemon controller #1518 would be the one replacing static pod?
I don't think it's a good idea to reference a feature that does not exist yet. Its name may change, it may be dropped, who knows how log will it take to merge it.

dchen1107 commented:
LGTM except above nits.

mikedanese commented:
@k8s-bot ok to test

mikedanese commented:
@dchen1107 I see that your nit has been addressed and I think this doc is generally useful for v1.1. I'm going to add LGTM and we can iterate on this if needed
